src/main/java/org/opensearch/security/resources/ResourceSharingIndexHandler.java

```diff
@@ -891,10 +936,190 @@ private void executeSearchRequest(
             Set<String> collectedResourceIds = new HashSet<>();
             String scrollId = initialResponse.getScrollId();
-            processScrollResults(collectedResourceIds, scroll, scrollId, initialResponse.getHits().getHits(), listener);
+            processScrollResultsAndCollectResourceIds(
+                collectedResourceIds,
+                scroll,
+                scrollId,
+                initialResponse.getHits().getHits(),
+                listener
+            );
         }, listener::onFailure);
     }
 
+    /**
+     * Executes a search request and returns a set of collected resource-sharing documents using scroll.
+     * @param resourceIndex the index whose records are to be searched
+     * @param scroll        Search scroll context
+     * @param searchRequest Initial search request
+     * @param query         Query builder for the request
+     * @param listener      Listener to receive the collected resource sharing records
+     */
+    private void executeAllSearchRequest(
+        String resourceIndex,
+        Scroll scroll,
+        SearchRequest searchRequest,
+        AbstractQueryBuilder<? extends AbstractQueryBuilder<?>> query,
+        ActionListener<Set<SharingRecord>> listener
+    ) {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder().query(query)
+            .size(1000)
+            .fetchSource(new String[] { "resource_id", "created_by", "share_with" }, null);
+
+        searchRequest.source(searchSourceBuilder);
+
+        StepListener<SearchResponse> searchStep = new StepListener<>();
+        client.search(searchRequest, searchStep);
+
+        searchStep.whenComplete(initialResponse -> {
+            Set<SharingRecord> recs = new HashSet<>();
+            String scrollId = initialResponse.getScrollId();
+            processScrollResultsAndCollectSharingRecords(
+                null,
+                true,
+                resourceIndex,
+                recs,
+                scroll,
+                scrollId,
+                initialResponse.getHits().getHits(),
+                listener
+            );
+        }, listener::onFailure);
+    }
+
+    /**
+     * Fetches resource-sharing records for this user for a given resource-index.
+     * Executes in 2 steps:
+     * Step-1:
+     *  - Fetch field mappings and create a terms query to match against all action-groups in shared_with block, as well as creator name.
+     * Step-2:
+     *  - Use the terms query to search resource-sharing index to fetch matching records.
+     *
+     * @param resourceIndex the index for which records are to be searched
+     * @param user the user that is requesting the records
+     * @param flatPrincipals user's name, roles, backend_roles to be used for matching.
+     * @param listener to collect and return accessible sharing records
+     */
+    @SuppressWarnings("unchecked")
+    public void fetchAccessibleResourceSharingRecords(
```

Comment:
Can this be re-written assuming that we will be able to obtain which resources are visible to a user given https://github.com/opensearch-project/security/pull/5600?

With that, you can directly search on the resource index (per page) get ids and then do an mget request for the respective sharing records.

Reply:
Makes sense. The mGet in a batch is an interesting approach. I had in mind that for each scroll it would issue an mGet so the batch for the mGet is the scroll window size. In any event, I don't expect existing dashboards use-cases to have thousands of "resoures" so what exists here is pretty defensive. In the future, it would be interesting to see what people build on top of dashboards with a better plugin developer experience.

Reply:
As discussed, the central page being built into dashboards is primarily to help with migration from backend-role based visbility check to full-block resource sharing and authz. This code path is the exceptional case and not the norm. 

When the `Share` button is built into existing dashboards plugins, then the primary operation on the resource sharing indices are GET operations given the resource type and id.